src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -136,11 +163,18 @@ struct LinkCollector<'a, 'tcx> {
     /// because `clean` and the disambiguator code expect them to be different.
     /// See the code for associated items on inherent impls for details.
     kind_side_channel: Cell<Option<(DefKind, DefId)>>,
+    /// Cache the resolved links so we can avoid resolving (and emitting errors for) the same link
+    visited_links: FxHashMap<CacheKey, Option<CachedLink>>,
```

Comment:
~~Hmm, won't this conflict with https://github.com/rust-lang/rust/pull/77700? How will the two interact?~~ never mind, looks like you just cherry-picked all those commits which is fine.

Reply:
I branched this off of #77700 so it's blocked on that, hence the draft PR.